### PR TITLE
tee-supplicant: tee_socket_recv(): report tructated datagrams

### PR DESCRIPTION
**Pull request by jenswi-linaro**
Updates tee_socket_recv() to detect and report truncated datagrams and
enable amount of data available to be queried.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
https://github.com/OP-TEE/optee_os/pull/4952 was just merged. This PR will also be needed before https://github.com/OP-TEE/optee_test/pull/555 can be merged.

File: tee-supplicant/src/tee_socket.c [hunk 1]
```diff
@@ -479,25 +493,91 @@ static TEEC_Result tee_socket_send(size_t num_params,
...
-	r = read(fd, buf, *blen);
+	if (*blen) {
+		/* If  *blen  ==  0, the timeout parameter has no effect. */
+		res = poll_with_timeout(&pfd, 1, timeout);
+		if (res != TEEC_SUCCESS)
+			return res;
+	}
+
+	if ((socktype == SOCK_DGRAM && *blen < SUPP_MAX_PEEK_LEN) || !*blen) {
+		/* Check how much data we have queued.  */
```
Comment (etienne-lms) [inline tee-supplicant/src/tee_socket.c:531]: double space chars

File: tee-supplicant/src/tee_socket.c [hunk 2]
```diff
@@ -479,25 +493,91 @@ static TEEC_Result tee_socket_send(size_t num_params,
...
 	if (r == -1) {
-		if (errno == EAGAIN || errno == EWOULDBLOCK || errno == EINTR)
-			return TEE_ISOCKET_ERROR_TIMEOUT;
-		return TEEC_ERROR_BAD_PARAMETERS;
+		e = errno;
+		goto err;
+	}
+	if (socktype == SOCK_DGRAM && (out_flags & MSG_TRUNC)) {
+		/*
+		 * The datagram have been truncated, return the best length
```
Comment (etienne-lms) [inline tee-supplicant/src/tee_socket.c:565]: s/has/have/

**Review by etienne-lms — COMMENTED** [review]
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` with minor comments.

**Comment by jenswi-linaro** [comment]
Comments addressed, tag applied.

**Comment by jforissier** [comment]
s/tructated/truncated/ ;)

**Comment by jenswi-linaro** [comment]
Commit message updated.

File: tee-supplicant/src/tee_socket.c [hunk 3]
```diff
@@ -479,25 +493,91 @@ static TEEC_Result tee_socket_send(size_t num_params,
 	return res;
 }
 
+static ssize_t recv_with_out_flags(int fd, void *buf, size_t len, int inflags,
+				   int *out_flags)
+{
+        struct iovec iov = { .iov_base = buf, .iov_len = len, };
+        struct msghdr msg = { .msg_iov = &iov, .msg_iovlen = 1, };
+	ssize_t r = recvmsg(fd, &msg, inflags);
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply tee-supplicant/src/tee_socket.c:501]: line 499 indents with space chars instead of a tab.

[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply tee-supplicant/src/tee_socket.c:501]: Thanks, I'll fix.

**Comment by jenswi-linaro** [comment]
Update

**Comment by jforissier** [comment]
`Acked-by: Jerome Forissier <jerome@forissier.org>`

File: tee-supplicant/src/tee_socket.c [hunk 4]
```diff
@@ -57,6 +57,20 @@
...
+ *
+ * For SOCK_DGRAM we try to be accurate up to 4096 bytes as
+ * that's our arbitrary chosen sensible upper size (with
+ * some margin). Larger size doesn't make much sense since
+ * anything larger than the MTU is bound to cause trouble
+ * on a congested network.
+ *
+ * For SOCK_STREAM we chose the same upper limit for
+ * simplicity. It doesn't matter if there's more queued,
+ * no data will be lost.
```
Comment (etienne-lms) [inline tee-supplicant/src/tee_socket.c:71]: Is this paragraph still applicable? It seems to me the previous sequence is unchanged regarding stream sockets.

Comment (jenswi-linaro) [reply tee-supplicant/src/tee_socket.c:71]: Note that this would be the `!*blen` case in `recv_with_out_flags()`, so it's still applicable.

Comment (etienne-lms) [reply tee-supplicant/src/tee_socket.c:71]: ok, thanks.

**Review by etienne-lms — APPROVED** [review]
My A-b tag stands (can replace with`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`).

**Comment by jenswi-linaro** [comment]
Squashed and tags applied.

**Pull request merged**